server/datastore/mysql/software.go

```diff
@@ -2877,301 +2838,536 @@ last_vpp_install AS (
-	if (opts.OnlyAvailableForInstall && !opts.VulnerableOnly) || (opts.IncludeAvailableForInstall && !opts.VulnerableOnly) {
-		namedArgs["vpp_apps_platforms"] = fleet.VPPAppsPlatforms
-		if fleet.IsLinux(host.Platform) {
-			namedArgs["host_compatible_platforms"] = fleet.HostLinuxOSs
-		} else {
-			namedArgs["host_compatible_platforms"] = []string{host.FleetPlatform()}
+		stmtAvailable, args, err := sqlx.Named(stmtAvailable, namedArgs)
+		if err != nil {
+			return nil, nil, err
+		}
+		stmtAvailable, args, err = sqlx.In(stmtAvailable, args...)
+		if err != nil {
+			return nil, nil, err
 		}
-		stmt += ` UNION ` + stmtAvailable
-	}
 
-	// must resolve the named bindings here, before adding the searchLike which
-	// uses standard placeholders.
-	stmt, args, err := sqlx.Named(stmt, namedArgs)
-	if err != nil {
-		return nil, nil, ctxerr.Wrap(ctx, err, "build named query for list host software")
-	}
-	stmt, args, err = sqlx.In(stmt, args...)
-	if err != nil {
-		return nil, nil, ctxerr.Wrap(ctx, err, "expand IN query for list host software")
+		stmtAvailable = ds.reader(ctx).Rebind(stmtAvailable)
+		var availableSoftwareTitles []*hostSoftware
+		err = sqlx.SelectContext(ctx, ds.reader(ctx), &availableSoftwareTitles, stmtAvailable, args...)
+		if err != nil {
+			return nil, nil, err
+		}
+		tempBySoftwareTitleID := make(map[uint]*hostSoftware, len(availableSoftwareTitles))
+		tmpByVPPAdamID := make(map[string]*hostSoftware, len(byVPPAdamID))
+		if opts.OnlyAvailableForInstall {
+			// drop in uninstalls as available
+			for _, s := range hostSoftwareUninstalls {
+				tempBySoftwareTitleID[s.ID] = s
+			}
+			// anything pending or failed is also "available"
+			for _, s := range hostSoftwareInstalls {
+				if *s.Status == "pending_install" || *s.Status == "failed_install" || *s.Status == "pending_uninstall" {
+					tempBySoftwareTitleID[s.ID] = s
+				}
+			}
+			// vpp apps pending to install or failed are also "available"
+			for _, s := range hostVPPInstalls {
+				if *s.Status == "pending_install" || *s.Status == "failed_install" {
+					tmpByVPPAdamID[*s.VPPAppAdamID] = s
+				}
+			}
+		}
+		for _, s := range availableSoftwareTitles {
+			// If it's a VPP app
+			if s.VPPAppAdamID != nil {
+				if _, ok := byVPPAdamID[*s.VPPAppAdamID]; !ok {
+					if opts.OnlyAvailableForInstall {
+						tmpByVPPAdamID[*s.VPPAppAdamID] = s
+					} else {
+						byVPPAdamID[*s.VPPAppAdamID] = s
+					}
+				}
+			} else {
+				// Ensure not to overwrite the records from software installs and software uninstalls
+				if _, ok := bySoftwareTitleID[s.ID]; !ok {
+					if opts.OnlyAvailableForInstall {
+						tempBySoftwareTitleID[s.ID] = s
+					} else {
+						bySoftwareTitleID[s.ID] = s
+					}
+				}
+			}
+		}
+		// Clear out all the previous software titles as we are only filtering for available software
+		if opts.OnlyAvailableForInstall {
+			bySoftwareTitleID = tempBySoftwareTitleID
+			byVPPAdamID = tmpByVPPAdamID
+		}
```

Comment:
I don't love this, because we ran the queries above and basically threw away some of the results. However, the statement available query is missing a ton of software.

Reply:
In particular (from lines 2871->2879):
```
			for _, s := range hostSoftwareUninstalls {
				tempBySoftwareTitleID[s.ID] = s
			}
			for _, s := range hostSoftwareInstalls {
				tempBySoftwareTitleID[s.ID] = s
			}
			for _, s := range hostVPPInstalls {
				tmpByVPPAdamID[*s.VPPAppAdamID] = s
			}
```
So anything that has been installed or uninstalled can be installed again.